### PR TITLE
Switch to foot

### LINKED ISSUE #82: What should be the default launcher?
[sway-launcher-desktop](https://github.com/Biont/sway-launcher-desktop):

- fast
- fits into the "cli-applications" style
- does "one thing only" (launching)
- not very popular

[wofi](https://hg.sr.ht/~scoopta/wofi):

- not being actively maintained
- does "one thing only" (launching) but allows to display custom content too
- bit slow

[bemenu](https://github.com/Cloudef/bemenu):

- never tried

I am very open to other proposals and will add them to this issue. Polling is open until 2021-08-11.

[](https://api.gh-polls.com/poll/01FAAJED1AA4HHR51MYT9WPH4G/sway-launcher-desktop/vote)
[](https://api.gh-polls.com/poll/01FAAJED1AA4HHR51MYT9WPH4G/wofi/vote)
[](https://api.gh-polls.com/poll/01FAAJED1AA4HHR51MYT9WPH4G/bemenu/vote)

### LINKED ISSUE #88: Terminal opens slowly
Hello! I've been following this project for a while and yesterday I decided to give it a try. One of the problems that I've faced is the time it takes from pressing Mod+Enter to seeing the terminal window - it's absurdly long for my machine (i7-8650U, 16GB RAM). 

I suppose the problem may be a rather slow script terminal.sh which checks OpenGL version EVERY time I open a terminal, which in Sway happens a lot more than in a conventional DE. This of course is unnecessary.
 
I have a suggestion to make it a one-time execution script, either in the installation session or when opening terminal for the first time. The other option would be to use a baseline terminal which is not dependent on graphics support, which makes sense if you think of Manjaro Sway not only as a Sway showcase, but as an option for low-end machines. 

May we have a poll or something along those lines about the default terminal emulator? For example, [foot](https://codeberg.org/dnkl/foot) may be a good fit, considering its' speed and lack of GPU acceleration, which is a pro in this context.

### PR DESCRIPTION
**Pull request by FallenChromium**
Resolves https://github.com/Manjaro-Sway/manjaro-sway/issues/88
Closes https://github.com/Manjaro-Sway/manjaro-sway/issues/82
It was a breeze to reconfigure the dotfiles to use foot, and I am very happy with the results (the terminal opens in an order of magnitude faster on my laptop now!). But it didn't go 100% smooth.
Currently this PR has two problems:
1. $TERM variable from ~/.profile has no effect, as it is changed when foot is launched. foot's terminfo is very new and using it may lead to serious compatibility troubles when ssh-ing to servers, for example. Foot has a flag `--term` to override the $TERM variable, shall we change $TERMINAL_COMMAND and $term variable in sway to use xterm-256color, to achieve better compatibility?
2. When wofi launches a CLI utility, it uses --exec flag with the `term` executable, foot doesn't have such flag. I tried to use a script to bypass that (just out of curiosity, personally I think that we shall switch from wofi asap), but I encountered a problem: there are three sources of truth for wofi to populate 'term' variable - from env $TERM, when wofi is launched with a --term flag or from a config file. It seems like env variable has the highest priority, because even with the config and --term flag overwritten to /usr/share/sway/scripts/terminal.sh, it still tried to open htop in foot. I need help on this one

I have tested the waybar on-click scripts, they are also opening way, way faster now, though I had to apply some fixes to the commands for them to work nicely with foot.
Looking forward for your feedback!

### CODE REVIEW

**Comment by dnkl** [comment]
Regarding 2)

It has been requested several times that we add `-e` to foot. See https://codeberg.org/dnkl/foot/issues/625 for the last one.

TL;DR `-e` isn't as portable as people think; even among terminals that do have the flag (see https://codeberg.org/dnkl/foot/issues/184#issuecomment-150624, and https://github.com/ranger/ranger/issues/1108#issuecomment-372113294), and thus we've discouraged its use.

However, I've also said that if there's "enough" reason, we can add it. It'd be a dummy flag. I.e. it would be ignored, but you'd at least be able to run things like `foot -e htop` without getting a parsing error. Up until now, people have been, mostly, happy with the explanation and haven't pursued things further.

That "offer" still stands. While I do think it would be better to patch applications to stop depending on `-e`, I would be ok with adding it for the sake of compatibility.

File: community/sway/etc/skel/.config/foot/foot.ini [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+# 16 = <256-color palette #16>
+# ...
+# 255 = <256-color palette #255>
+# selection-foreground=<inverse foreground/background>
+# selection-background=<inverse foreground/background>
+# jump-labels=<regular0> <regular3>
+# urls=<regular3>
+
+[csd]
+# preferred=server
```
Comment (boredland) [inline community/sway/etc/skel/.config/foot/foot.ini:91]: shouldn't we disable csd?
```suggestion
preferred=none
```

Comment (FallenChromium) [reply community/sway/etc/skel/.config/foot/foot.ini:91]: Hm, I don't see any decorations even with the default config, probably because gtk3-nocsd is the default gtk3 provider for Manjaro Sway. If it affects anything else then sure, why not!

**Comment by boredland** [comment]
I think this could be a good reason to first migrate away from wofi. I just tested if `TERMINAL_COMMAND=foot sway-launcher-desktop` launches htop without further ado. It does.

**Comment by FallenChromium** [comment]
> Regarding 2)

Nice to see you there, thanks for the great terminal :)

> However, I've also said that if there's "enough" reason, we can add it.

I don't see wofi being hardcoded to use --exec flag as a good reason to encourage this behavior in terminals, I think we will switch to different launcher to solve this issue.

File: community/sway/etc/skel/.config/foot/foot.ini (hunk 1 above)
Comment (dnkl) [reply community/sway/etc/skel/.config/foot/foot.ini:91]: `preferred` is just that - we ask the compositor very nicely to use either SSD (`preferred=server`), or CSD (`preferred=client`). The compositor may choose to override that. Sway for example, does.

Furthermore, this setting is completely ignored on compositors that doesn't implement the _XDG toplevel decoration_ Wayland protocol (only GNOME/mutter, that I'm aware of).

`preferred=none` on the other hand will disable **both** SSD and CSDs (assuming a floating window in Sway - tiled and stacked windows get SSDs anyway).

**Comment by FallenChromium** [comment]
The latest commit leads to absolutely crazy (= FAST) response time for floating windows (which are used for nmtui, htop and sensors), they open almost instantly. Big thanks to dnkl's server-client feature where `foot --server` has to be launched once to load glyphs\fonts and do other time-consuming operations and then `footclient` can be launched on demand to use that performance boost. I was thinking about making every terminal a client to initial foot --server process, but because it's said in the foot's wiki that it can lead to issues if one of the windows is heavily I\O-loaded and, if the server process will crash (though in my practice I've never seen that), all terminal windows will be gone. 

I am being conservative here so I decided to leave the default terminal window to be standalone, foot's footprint (see what I did right there? :D) isn't big anyways, but overriding $term to `footclient` can be left as a suggestion for low-end PC users. Soo... Anything else blocking this PR? I think we only have to decide on first point in the first message, are we going to use xterm-256color terminfo by default to maintain ssh compatibility etc., or we will stay with foot's terminfo as default? I'm excited to let it go live!

File: community/sway/etc/skel/.config/foot/foot.ini (hunk 1 above)
Comment (boredland) [reply community/sway/etc/skel/.config/foot/foot.ini:91]: Ah, thx for clarification!

**Comment by boredland** [comment]
what would be the implication setting it to `xterm-256color` other than being less annoying when ssh-ing?

**Comment by dnkl** [comment]
> what would be the implication setting it to xterm-256color other than being less annoying when ssh-ing?

There are two implications.

One, foot's own terminfo has a couple things not in the xterm one: `bw`, `Sync`, `setrgbf`, `setrgbb` and a different `flash` are the ones I can think of right now. `Sync` is probably the most interesting one; it tells applications (tmux, for example) that the terminal supports synchronized updates (avoids tearing-like rendering).

Two, while foot supports all escapes **currently** in use by xterm's terminfo, future updates may change that, resulting in a broken terminal. This has happened in the past: when xterm added `REP`, terminals that defaulted to `xterm-256color` broke in at least ncurses-based applications.

File: community/sway/etc/skel/.profile [hunk 2]
```diff
@@ -15,9 +15,9 @@ export QT_WAYLAND_DISABLE_WINDOWDECORATION="1"
 
 # set default shell and terminal
 export SHELL=/usr/bin/zsh
-export TERM=xterm-kitty
-export TERMINAL_COMMAND='/usr/share/sway/scripts/terminal.sh'
+export TERM=xterm-256color
```
Comment (boredland) [inline community/sway/etc/skel/.profile]:
```suggestion
```

File: community/sway/etc/sway/config.d/99-autostart-applications.conf [hunk 3]
```diff
@@ -12,7 +12,7 @@ exec {
     '[ -x "$(command -v wlsunset)" ] && /usr/share/sway/scripts/sunset.sh'
     '[ -x "$(command -v autotiling)" ] && autotiling'
     '[ -x "$(command -v sworkstyle)" ] && sworkstyle &> /tmp/sworkstyle.log'
-}
+    $term_server
```
Comment (boredland) [inline community/sway/etc/sway/config.d/99-autostart-applications.conf]: what is going on here?
```suggestion
    '[ -x "$(command -v foot)" ] && foot --server'
}
```

File: community/sway/etc/sway/definitions [hunk 4]
```diff
@@ -24,9 +24,11 @@ set $background /usr/share/backgrounds/manjaro-sway/gruvbox-dark-manjarosway-384
 set $theme /usr/share/sway/themes/matcha-green
 
 # Your preferred terminal emulator
-set $term /usr/share/sway/scripts/terminal.sh
-set $term_float $term float
-set $term_float_portrait $term float_portrait
+set $term foot
+set $term_client footclient
```
Comment (boredland) [inline community/sway/etc/sway/definitions]: is there a reason to pack that into a variable?
```suggestion
# make a foot client instance the default terminal for even faster terminal windows
# set $term footclient
```

File: community/sway/etc/sway/config.d/99-autostart-applications.conf (hunk 3 above)
Comment (FallenChromium) [reply community/sway/etc/sway/config.d/99-autostart-applications.conf]: It is to launch the $term_server command defined in /etc/sway/definitions. Still not sure what's the most correct way to do this, probably can be rewritten to static `[ -x "$(command -v foot)" ] && foot --server`, if that would be better in your opinion, I can do this, now I realized it's over-engineering to add it. If a user will change to other client-sever terminal, $term_float shall be redefined anyway and a new server process has to be added to the ~/.config/sway/config.d/00-user.conf (or by changing $term_server, in the current variant). I'll change it I think

File: community/sway/etc/sway/definitions (hunk 4 above)
Comment (FallenChromium) [reply community/sway/etc/sway/definitions]: It's for launching floating windows with $term_client. Though my current setup is to statically use 'footclient -a floating_shell', I just wasn't sure if it's good enough to leave it static, so I made this solution, which I don't feel is great either.

File: community/sway/etc/sway/definitions [hunk 5]
```diff
@@ -24,9 +24,11 @@ set $background /usr/share/backgrounds/manjaro-sway/gruvbox-dark-manjarosway-384
...
+set $term_server foot --server
```
Comment (boredland) [inline community/sway/etc/sway/definitions]:
```suggestion
```
i'd prefer this not to be a variable that is only being used once, because I don't see a scenario where sb would want to override this. and instead will only complicate the readability of the autostart config.

File: community/sway/etc/skel/.profile (hunk 2 above)
Comment (FallenChromium) [reply community/sway/etc/skel/.profile]: Yep, isn't working anyway. What's about changing $term to `foot --term xterm-256color`, are we doing it by default or we're leaving solving ssh issues to the users?

File: community/sway/etc/sway/config.d/99-autostart-applications.conf (hunk 3 above)
Comment (boredland) [reply community/sway/etc/sway/config.d/99-autostart-applications.conf]: sorry. updated and makes sense.  with whats going on I meant the '}' being removed

File: community/sway/etc/sway/definitions (hunk 4 above)
Comment (FallenChromium) [reply community/sway/etc/sway/definitions]: It would be wise to move the suggestion that you made to the etc/skel/.config/sway/config.d/00-user.conf for example, so it will be easier, just uncomment it and you're done. Probably adding a word for low-memory footprint or\and potential data loss if foot --server crashes would be nice, too

File: community/sway/etc/sway/definitions (hunk 5 above)
Comment (FallenChromium) [reply community/sway/etc/sway/definitions]: Agreed

File: community/sway/etc/skel/.profile (hunk 2 above)
Comment (boredland) [reply community/sway/etc/skel/.profile]: didn't try it, but shouldn't adding 
```
alias ssh="TERM=xterm-256color ssh"
```
to the .zshrc be working? if so I'd prefer that

File: community/sway/etc/sway/definitions (hunk 4 above)
Comment (FallenChromium) [reply community/sway/etc/sway/definitions]: So just to clarify: if we're getting rid of $term_client, are we going to replace $term_float to static `footclient` command, or we're using $term here to simplify configuration of third-party terminal? Because then these optimizations for floating windows are lost, unfortunately

File: community/sway/etc/skel/.profile (hunk 2 above)
Comment (FallenChromium) [reply community/sway/etc/skel/.profile]: That's a valid point (actually that's how I do it on my machine), but I fear that it can break things if user will install and use other terminal by default. It seems like a good solution, but we should somehow inform users that we made this. At this point it will lead to a giant readme, time to make a wiki? 😆

File: community/sway/etc/sway/definitions (hunk 4 above)
Comment (boredland) [reply community/sway/etc/sway/definitions]: You're absolutely right. Let's just keep it like it is now and add a nice hint to the 00-user.conf on setting $term to $term_client for warp speed.

File: community/sway/etc/skel/.profile (hunk 2 above)
Comment (boredland) [reply community/sway/etc/skel/.profile]: If a user changes shells, I expect him to know what he's doing... Regarding wiki: we could actually use github wiki or github docs for that in the manjaro-sway repo. Perhaps not a full fledged wiki, but a 'faq' page, perhaps searchable..

Comment (boredland) [reply community/sway/etc/skel/.profile]: I mean: if a user breaks it, he will google and find a solution in no time.

Comment (FallenChromium) [reply community/sway/etc/skel/.profile]: Makes sense to me. Then adding the mentioned line to .zshrc is a good middleground. Would you like to add it as a suggestion? I can do it myself, but, you know, I just kinda want to show that this PR was not made by me only 😄

**Comment by fraunos** [comment]
Hey, as a user - what should I do to update Manjaro Sway to get these updated configs, so I can use preconfigured foot terminal?

**Comment by boredland** [comment]
`pacman -S manjaro-sway/manjaro-sway-settings`
https://github.com/Manjaro-Sway/manjaro-sway/blob/main/SUPPORT.md#how-can-i-update-the-manjaro-sway-settings-package-and-all-other-packages-from-manjaro-sway
